transformer_engine/pytorch/csrc/extensions/softmax.cu

```diff
@@ -245,3 +247,357 @@ at::Tensor scaled_aligned_causal_masked_softmax_backward(at::Tensor output_grad_
+
+  size_t cur_vocab_parallel_logits_ptr_begin = rowIdx * n_dim;  // 0, 1025
+  size_t cur_vocab_parallel_logits_ptr_end =
+      rowIdx * n_dim + n_dim;  //cur_vocab_parallel_logits_ptr_end = 1025, 2050
+
+  size_t end_mol_num = cur_vocab_parallel_logits_ptr_end % 8;  //end_mol_num = 1, end_mol_num = 2
+  size_t begin_mol_num = n_dim - end_mol_num;  //begin_mol_num = 1024, begin_mol_num = 1023
+
+  //valid range for evry row is [begin_offset, end_offset]
+  size_t begin_offset = begin_mol_num % 8;      //begin_offset = 0, begin_offset = 7
+  size_t end_offset = n_dim - end_mol_num - 1;  //end_offset = 1023, end_offset = 1022
+
+  float cur_row_exp_sum = sum_exp_logits_ptr[rowIdx];
+  float cur_row_max = logits_max_ptr[rowIdx];
+  float row_item = 0;
+  float cur_thread_softmax_log_mean = 0.0;
+
+  typedef cub::BlockReduce<float, BlockSize> BlockReduceT;
+  __shared__ typename BlockReduceT::TempStorage temp_storage;
+
+  for (size_t i = begin_offset + tid * 8; i <= end_offset - 7; i += 8 * BlockSize) {
+    int4 int4_arr = *reinterpret_cast<int4*>(
+        &vocab_parallel_logits_ptr[cur_vocab_parallel_logits_ptr_begin + i]);
+    dtype* bf_16_p = reinterpret_cast<dtype*>(&int4_arr);
+#pragma unroll
+    for (int k = 0; k < 8; k++) {
+      dtype data_bf16 = bf_16_p[k];
+      float data_fp32 = float(data_bf16);  //convert to float
+      row_item = compute_mean_log(data_fp32, cur_row_max, cur_row_exp_sum);
+      cur_thread_softmax_log_mean += row_item;  //sum all "log value"
+    }
+  }
+
+  float row_log_sum = BlockReduceT(temp_storage).Sum(cur_thread_softmax_log_mean);
+
+  if (threadIdx.x == 0) {
+#pragma unroll
+    for (size_t k = cur_vocab_parallel_logits_ptr_begin;
+         k < (cur_vocab_parallel_logits_ptr_begin + begin_offset); k++) {
+      float val = float(vocab_parallel_logits_ptr[k]);
+      row_item = compute_mean_log(val, cur_row_max, cur_row_exp_sum);
+      row_log_sum += row_item;
+    }
+#pragma unroll
+    for (size_t k = cur_vocab_parallel_logits_ptr_begin + end_offset + 1;
+         k < cur_vocab_parallel_logits_ptr_end; k++) {
+      float val = float(vocab_parallel_logits_ptr[k]);
+      row_item = compute_mean_log(val, cur_row_max, cur_row_exp_sum);
+      row_log_sum += row_item;
+    }
+    mean_log_probs_ptr[rowIdx] = row_log_sum / n_dim;
+  }
+}
+
+float __device__ __forceinline__ compute_exp_bwd_smooth(float row, float logits_max,
+                                                        float sum_exp_logits, size_t i,
+                                                        long masked_target_1d, float softmax_update,
+                                                        float label_smoothing, float smoothing,
+                                                        float average_grad, float grad_output) {
+  row = exp(row - logits_max);
+  row /= sum_exp_logits;
+  if (i == (size_t)masked_target_1d) {  // i == masked_target_1d
+    row = row - softmax_update;
+  }
+  if (label_smoothing > 0) {
```

Comment:
Why not just use smoothing here - you waste a register by holding onto this value.

Reply:
I'm sorry I didn't understand what you meant. Could you explain it specifically? THX @ptrendx

Reply:
You pass both `label_smoothing` and `smoothing` to this function, whereas smoothing value seems to be directly tied to label_smoothing value via
```cpp
smoothing = label_smoothing * vocab_size / (vocab_size - 1);
```
formula. The condition `label_smoothing > 0` is equivalent to `smoothing > 0`, right?  So you could save a register by just forgetting the `label_smoothing` value after the initial setting of smoothing (or maybe even just pass the value of `smoothing` to the kernel in the first place to not perform this computation by all the threads) and just use `smoothing` instead:
```cpp
if (smoothing > 0) {
  row -= smoothing * average_grad;
}
```
(considering that it seems `smoothing` can be only either 0 or >0 you could even just remove the conditional completely and always do `row -= smoothing * average_grad;`).